### PR TITLE
Display progress indicator while exporting screenshot

### PR DESCRIPTION
**Pull request by ssoloff**
This PR addresses half of issue #982.  Namely, it provides functionality to display a progress indicator while exporting a screenshot.

There are several design choices that need to be reviewed by the maintainers before this PR can be merged.  I will address each of these inline by annotating the changes.  Therefore, please hold off on reviewing this PR until I post a follow-up comment indicating the annotations are complete.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -6,6 +6,9 @@ addons:
 jdk:
 - oraclejdk8
 install: true
+before_script:
```
Comment (ssoloff) [inline .travis.yml]: This is necessary to allow the new integration tests added in this PR to run on Travis.  The integration tests spin up a new EDT, thus there needs to be a virtual display in order to avoid a `HeadlessException`.

File: src/games/strategy/triplea/ui/MapPanel.java [hunk 2]
```diff
@@ -462,10 +462,8 @@ public void setTopLeft(final int x, final int y) {
   }
 
   // this one is useful for screenshots
-  @Override
-  public void print(final Graphics g) {
+  public void drawMapImage(final Graphics g) {
     final Graphics2D g2d = (Graphics2D) g;
-    super.print(g2d);
```
Comment (ssoloff) [inline src/games/strategy/triplea/ui/MapPanel.java]: In order to display a progress indicator, I had to move the core logic for saving a screenshot to a background thread.  Of all the code in this method, the only bit that appeared to require running on the EDT was the call to `super.print()`.  I compared screenshots generated with this line both present and removed, and both screenshots were identical.  Therefore, I concluded this call had no effect on the final image and could be safely removed.

I renamed the method, not only because I removed the `super.print()` call, but because I didn't want to imply the method must be called from the EDT.  **Please advise if the method name should be reverted or changed to something else.**

In retrospect, adding Javadocs for this method, including a statement that the method is safe to call from a non-EDT thread, would probably be a good idea.

File: src/games/strategy/triplea/ui/menubar/ExportMenu.java [hunk 3]
```diff
@@ -128,7 +133,7 @@ private void exportXMLFile() {
 
   private void addSaveScreenshot(final JMenu parentMenu) {
     final AbstractAction abstractAction = SwingAction.of("Export Map Snapshot", e -> {
-
+      // get current history node. if we are in history view, get the selected node.
```
Comment (ssoloff) [inline src/games/strategy/triplea/ui/menubar/ExportMenu.java]: This comment was originally in `saveScreenshot()` and appeared to have been orphaned there when the following code was moved to this location at some point in the past.  I moved it simply to avoid confusion in `saveScreenshot()`.

File: src/games/strategy/triplea/ui/menubar/ExportMenu.java [hunk 4]
```diff
@@ -168,27 +173,48 @@ public String getDescription() {
...
           return;
         }
       }
+
       final File file = f;
-      final Runnable t = () -> {
-        if (saveScreenshot(node, file, frame, gameData)) {
-          JOptionPane.showMessageDialog(null, "Map Snapshot Saved", "Map Snapshot Saved",
-              JOptionPane.INFORMATION_MESSAGE);
+      final SwingWorker<?, ?> task = new SwingWorker<Void, Void>() {
```
Comment (ssoloff) [inline src/games/strategy/triplea/ui/menubar/ExportMenu.java]: This is the core of the change.  It runs the actual export on a background thread, displays a progress indicator,  blocks the user from interacting with the UI while the operation is running, and displays a success/failure message after the operation is complete.

There seems to be a lot going on across the two `saveScreenshot()` overloads that's unrelated to building the `ExportMenu`.  Therefore, I was tempted to extract a new class to isolate this functionality.  (Note also that the public `saveScreenshot()` method is called from outside this class by the `TripleAFrame` class.)  I held off because of my lack of knowledge of the TripleA architecture.

**Please advise if the public `saveScreenshot()` method should be extracted to a new class.**  If so, please suggest an appropriate package and class name.  I had previously considered `g.s.triplea.ui.export.ScreenshotExporter` based on what I saw related to the engine data exporters, but that may not be appropriate in this situation.

**Comment by ssoloff** [comment]
I've annotated everything I wanted to.  Please review whenever it's convenient.

**Comment by DanVanAtta** [comment]
Thank you for the annotations, I should be able to take a look later tonight (about 10 hours from now)

File: src/games/strategy/triplea/ui/menubar/ExportMenu.java (hunk 4 above)
Comment (DanVanAtta) [inline src/games/strategy/triplea/ui/menubar/ExportMenu.java]: IMO first extract the SwingWorker + progress bar to be generic. `SwingComponents.java` is a default place for swing component factory methods.

Then you could do something like:
- have Screenshot be defined in a class that implements Runnable
- you could then write something like `SwingComponents.runWithProgressBar("message", SaveScreenshot::new)`

Comment (ssoloff) [inline src/games/strategy/triplea/ui/menubar/ExportMenu.java]: Sounds good.  I took a brief look at `SwingComponents` while working this PR for an unrelated reason and didn't make the connection for it being a good place to extract a generic progress runner.

@DanVanAtta Did you envision `SaveScreenshot` (and all the code related to it) still being internal to `ExportMenu`?  Or would it be preferable to extract it to its own file in a different package?  If external, can you suggest an appropriate package, please?

File: src/games/strategy/ui/ProgressDialog.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+   * Initializes a new instance of the {@code ProgressDialog} class.
+   *
+   * @param owner The {@code Frame} from which the dialog is displayed or {@code null} to use a shared, hidden frame as
+   *        the owner of the dialog.
+   * @param message The progress message; must not be {@code null}.
+   */
+  public ProgressDialog(final Frame owner, final String message) {
+    super(owner, true);
+
+    Objects.requireNonNull(message);
```
Comment (ssoloff) [inline src/games/strategy/ui/ProgressDialog.java]: Chalk that up to laziness.  I didn't even bother to look at the rest of the codebase to see if there were existing libraries being used that provide this functionality.  (I probably would have if it didn't exist in the Java standard library.)  I'll switch to Guava as needed.

File: src/games/strategy/triplea/ui/menubar/ExportMenu.java [hunk 6]
```diff
@@ -246,20 +272,13 @@ private static boolean saveScreenshot(final HistoryNode node, final File file, f
...
 
       // save Image as .png
-      try {
-        ImageIO.write(mapImage, "png", file);
-      } catch (final Exception e2) {
-        e2.printStackTrace();
-        JOptionPane.showMessageDialog(frame, e2.getMessage(), "Error saving Screenshot", JOptionPane.OK_OPTION);
-        retval = false;
-      }
+      ImageIO.write(mapImage, "png", file);
```
Comment (ssoloff) [inline src/games/strategy/triplea/ui/menubar/ExportMenu.java]: The error handling was moved up to the public `saveScreenshot()` overload.  I modified the private overload (which runs on a non-EDT thread, and thus shouldn't display a message dialog) to throw an exception on failure.  That exception is now handled in the public overload (which runs on the EDT), and it is responsible for displaying both the success and failure message dialogs.

File: src/games/strategy/triplea/ui/menubar/ExportMenu.java (hunk 4 above)
Comment (DanVanAtta) [inline src/games/strategy/triplea/ui/menubar/ExportMenu.java]: To its own place. I think that after looking at the test case and seeing that we have to test UI logic to get to this component. To fix that test, testing the screenshot logic directly, using an interface for the UI component interactions I believe would 'fix' the test problem, so that we have a relatively clean isolated test. The UI testing then IMO can be done by hand, in theory it's all library code anyways.

File: src/games/strategy/ui/ProgressDialog.java (hunk 5 above)
Comment (DanVanAtta) [inline src/games/strategy/ui/ProgressDialog.java]: Caution when there are Guava exact equivalents to what is available now in Java8. For example, `Sets.newHashSet()` in Guava is deprecated now (it was there back when JDK6 or 7 were out, before empty diamon notation, ie: `new HashSet<>()`)

In those cases we should be using the java version. In this case though `checkNotNull` has additional functionality.

**Comment by ssoloff** [comment]
@DanVanAtta In response to your general feedback:

- I didn't really look into consolidating progress bar behavior across different areas of the application.  Honestly, I didn't tackle this issue because it was a feature I was particularly interested in seeing in the game.  Rather, I was browsing the open issues looking for some low-hanging fruit in which I could get comfortable with the TripleA code.
- I totally agree with your philosophy on testing.  I had considered doing something similar to what you suggested but felt it might be construed as invasive.  As this is my first significant contribution to TripleA, I didn't want to "perturb the pendulum" too far from equilibrium. :smiley:  Given how `SwingWorkerCompletionWaiter` was developing in this PR, I felt a pure unit test (by mocking the `SwingWorker` behavior) wouldn't have provided much value.  Hence, why I went with an integration test to verify my understanding of the actual `SwingWorker` behavior at runtime (which turned out to be a bit more asynchronous than I expected).  When I address your recommendation to generify the progress runner stuff, I'll review the tests once more while keeping your concerns in mind.  I'm sure we'll be able to reach a consensus in one more iteration after I submit those changes.
- I'll reorder methods per the guidelines you provided.
- As I alluded to in the first bullet, my experience with TripleA is minimal, so I don't have much experience with the various places progress bars appear or should appear.

File: src/games/strategy/triplea/ui/menubar/ExportMenu.java [hunk 7]
```diff
@@ -168,27 +173,48 @@ public String getDescription() {
...
+              JOptionPane.showMessageDialog(frame, "Map Screenshot Saved", "Map Screenshot Saved",
+                  JOptionPane.INFORMATION_MESSAGE);
+            });
+          } catch (final ExecutionException e) {
+            SwingUtilities.invokeLater(() -> {
+              JOptionPane.showMessageDialog(frame, e.getCause().getMessage(), "Error Saving Screenshot",
+                  JOptionPane.OK_OPTION);
+            });
+          } catch (final InterruptedException e) {
+            Thread.currentThread().interrupt();
```
Comment (DanVanAtta) [inline src/games/strategy/triplea/ui/menubar/ExportMenu.java]: Interesting read, sounds like it advocates rethrowing the interrupt.

My reading/understanding of how to handle is to abort/stop any on-going operations, clean-up and return. The JVM has received a terminate signal, so things 'should' be shutting down.  

Rather than be consistent in handling, I'd prefer to discuss what the better handling is, and try to create a library function that we can apply consistently. The library function will then make this a no-brainer of an operation, which is good for going over and writing lots of code : )

- does seem like interrupting the current thread manually is not needed. I recall reading it's not the most predictable thing. 
- we could rethrow the interrupted exception. Wrapping it probably is called for so that we don't have to have checked exceptions in the entire stack. Alternatively we do a 'return' from method.

Comment (ssoloff) [inline src/games/strategy/triplea/ui/menubar/ExportMenu.java]: I believe Goetz's gist is

- Allow `InterruptedException` to propagate up the call stack, if possible.
- If it's not possible to propagate the `InterruptedException`, re-interrupt the thread so the fact that the thread has been interrupted will not be lost.

I think these recommendations go back to how Goetz et al. wanted a canonical mechanism to cancel background tasks when they were designing `java.util.concurrent` many years ago.

Note that wrapping an `InterruptedException` in an unchecked exception would not work for code (e.g. `Executor` implementations) that might be acting on the thread interruption status, as the mere act of catching an `InterruptedException` (and not re-throwing it as an `InterruptedException`) clears the thread's interruption status (hence, why Goetz suggests re-interrupting the thread).

I agree with your idea for introducing a library function to encapsulate the preferred behavior.  When you've decided what that behavior should be, let me know, and I'll extract that as part of this PR, if you like.

Comment (DanVanAtta) [inline src/games/strategy/triplea/ui/menubar/ExportMenu.java]: IMO to get the most consistent behavior, log the interrupted exception and return. Add a TODO with a link to the Goetz article, we very well may want to consider interrupting the current thread. If we really want to follow up a github issue would be a good way to track it. This should allow you to proceed while we research differences in system behavior on interrupt vs not.

**Comment by ssoloff** [comment]
@DanVanAtta I've pushed a new batch of commits addressing the issues raised in the past 24 hours.  I've summarized the changes below.

First, the easy stuff:

- `Objects.requireNotNull` replaced with `Preconditions.checkNotNull`.
- Reorder methods per previously referenced project code guidelines.
- Remove call to `Thread.currentThread().interrupt()` and replace with TODO (I'll create an issue once the PR is merged).

Now for the stuff that may require another round of review:

- I extracted the method `SwingComponents.runWithProgressBar()` as you suggested.  However, I'm not sure the design I chose is the best option.  I opted for a promise-based approach, but I could have just as easily required the user to specify success and failure callbacks and return the actual `SwingWorker`.  The current implementation isn't fully functional.  For example, it does not support canceling the background task (simply because the only consumer, the export screenshot command, does not require it).  My thought at this point is to wait until someone requires this feature before adding it.  **Please review how it's used and let me know if you had something else in mind.**
- Extracting a new class to export the screenshot didn't go exactly as I thought it would when I read your most recent comment on this issue.  I was not able to simply pass a single `Runnable` or `Callable` to `runWithProgressBar()` from the top-level callers (`TripleAFrame` and `ExportMenu`) because the operation really has two parts: a UI part to prompt the user for the file and a non-UI part to generate and write the screenshot image; only the non-UI part should run within the context of `runWithProgressBar()`.  I made the call to go with something a bit more vanilla, although I did refactor it somewhat so it is more than simply a new static method on a new class.  **Please review and let me know what changes would align it more with your original thinking.**

**Comment by Cernelius** [comment]
Why talking about screenshot instead of snapshot?

Since now in the menu it has been changed to the more correct "Snapshot" (I would have preferred something like "Export Game Image", but surely much better than "Screenshot"), why not changing it so also for all occurrences in the engine? It would also reduce confusion about what you are talking about, because now I guess somebody may wonder if you are talking about the snapshot, when you are saying screenshot.

"Export Screenshot" would mean that it is exporting an image of what you are seeing on the screen only, and may be another item, if something wants to add it, beside the snapshot function.

Off topic but similar, after 1.8.0.9 I've lost the display progress indicator while loading an xml, which is particularly bad for games like War of the Relics, as it looks like the game just didn't open, and you have to wonder if it didn't or it is loading.

**Comment by ssoloff** [comment]
@Cernelius I used the term *screenshot* in this PR for the following reasons:

- the original issue (#982) used this term, and
- the code itself uses this term for various method and field names.

Unfortunately, I was not aware of the discussion that occurred in #1226 (which is what I believe you are referring to) and so assumed that  *screenshot* was the preferred terminology for this PR.  I'll leave it up to @DanVanAtta to decide if the code should be refactored to use _snapshot_ instead of _screenshot_ in identifiers as part of this PR.

Note also that _screenshot_ is still used in the UI.  Specifically, in the Game History panel context menu, the command label is **Save Screenshot** (but it runs the exact same code as the main menu's **Export > Export Map Snapshot** command).  It's possible that this label was simply overlooked as part of 0e86423.  @DanVanAtta or @ron-murhammer, please advise if that label should be updated to use _snapshot_ as part of this PR.

**Comment by Cernelius** [comment]
> in the Game History panel context menu, the command label is Save Screenshot

Yeah, that should be changed too.

**Comment by DanVanAtta** [comment]
Re: snapshot vs screenshot

We have in the code as an internal name consistently "screenshot" (instance counts):
```
dan@dan-desk:~/work/triplea$ find . -name "*.java" -type f | xargs grep -ir "screenshot" | wc -l
21
dan@dan-desk:~/work/triplea$ find . -name "*.java" -type f | xargs grep -ir "snapshot" | wc -l
2
```

- The internal naming should stay consistent. 
- It appears the external naming is not consistent, I would agree it should be updated to consistently read  'export map snapshot'. But, that update is not in scope for this branch and PR. Such a change can easily happen in its own branch and respective PR.
- If we want to update the internal naming to 'snapshot', or perhaps better `mapSnapshot`, that is okay. But again that should likely be in its own branch.

File: .travis.yml (hunk 1 above)
Comment (ssoloff) [inline .travis.yml]: The integration test that requires this capability is for the new `SwingWorkerCompletionWaiter` class, which _is_ pure Swing code.  However, to avoid introducing this requirement, I will change the integration test to a unit test that uses a test double for `SwingWorker`.

File: src/games/strategy/ui/SwingComponents.java [hunk 8]
```diff
@@ -278,4 +285,50 @@ public static JMenu newJMenu(final String menuTitle, final KeyboardCode keyboard
...
+   * @param frame The {@code Frame} from which the progress dialog is displayed or {@code null} to use a shared, hidden
+   *        frame as
+   *        the owner of the progress dialog.
+   * @param message The message to display in the progress dialog; must not be {@code null}.
+   * @param task The task to be executed; must not be {@code null}.
+   *
+   * @return A promise that resolves to the result of the task; never {@code null}.
+   */
+  public static <T> CompletableFuture<T> runWithProgressBar(
+      final Frame frame,
```
Comment (ssoloff) [inline src/games/strategy/ui/SwingComponents.java]: My bad.  I keep forgetting that the TripleA Eclipse formatter configuration does not join comment lines.  Will fix.

File: src/games/strategy/triplea/ui/export/ScreenshotExporter.java [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+import games.strategy.engine.history.Round;
+import games.strategy.triplea.ui.IUIContext;
+import games.strategy.triplea.ui.MapPanel;
+import games.strategy.triplea.ui.TripleAFrame;
+import games.strategy.triplea.ui.mapdata.MapData;
+import games.strategy.ui.SwingComponents;
+import games.strategy.ui.Util;
+
+/**
+ * Exports a game screenshot.
```
Comment (ssoloff) [inline src/games/strategy/triplea/ui/export/ScreenshotExporter.java]: Agreed.  Will remove.

File: src/games/strategy/triplea/ui/export/ScreenshotExporter.java [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+   * at the specified history step to that file.
+   * 
+   * @param frame The frame associated with the game screenshot to export; must not be {@code null}.
+   * @param gameData The game data; must not be {@code null}.
+   * @param node The history step at which the game screenshot is to be captured; must not be {@code null}.
+   * 
+   * @return A promise that is fulfilled when the save screenshot operation is complete or empty if the user aborted the
+   *         save; never {@code null}.
+   */
+  public static Optional<CompletableFuture<Void>> promptForFileAndRunSave(final TripleAFrame frame,
```
Comment (ssoloff) [inline src/games/strategy/triplea/ui/export/ScreenshotExporter.java]: Agreed.  This was a YAGNI anticipating that someone would need the promise in the future.  Will change to `void`.

File: src/games/strategy/triplea/ui/export/ScreenshotExporter.java [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+      final GameData gameData, final HistoryNode node) {
+    final ScreenshotExporter exporter = new ScreenshotExporter(frame);
+    return exporter.promptForFile()
```
Comment (ssoloff) [inline src/games/strategy/triplea/ui/export/ScreenshotExporter.java]: Agreed based on the previous comment to change the return value to `void`.  Will adjust as suggested.

File: src/games/strategy/triplea/ui/export/ScreenshotExporter.java [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+        .map(file -> Optional.of(exporter.runSave(gameData, node, file)))
+        .orElse(Optional.empty());
+  }
+
+  /**
+   * Prompts the user for the file to which the screenshot will be saved.
+   * 
+   * @return The file to which the screenshot will be saved or empty if the user aborted the save; never {@code null}.
+   */
+  public Optional<File> promptForFile() {
```
Comment (ssoloff) [inline src/games/strategy/triplea/ui/export/ScreenshotExporter.java]: I like the idea of extracting a new method to `SwingComponents` to reuse some of the file prompting behavior, but please note that the legacy code now contained within `promptForFile()` has some very specific behavior related to file extensions that may get out of control when they are generified.

I'll put together "the simplest thing that works" for the current use case so as not to unnecessarily increase the scope of this PR.  I'll let you decide if it's too specific. 😃 

BTW, related to #1423: the file filter in the current implementation has a description of "Saved Screenshots".  Do you want me to leave that unchanged in this PR and just add a comment to #1423 that it should be addressed there?

File: src/games/strategy/triplea/ui/export/ScreenshotExporter.java [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+  /**
+   * Saves the screenshot for the specified game at the specified history step to the specified file.
+   * 
+   * @param gameData The game data; must not be {@code null}.
+   * @param node The history step at which the game screenshot is to be captured; must not be {@code null}.
+   * @param file The file that will receive the screenshot; must not be {@code null}.
+   * 
+   * @return A promise that is fulfilled when the save screenshot operation is complete; never {@code null}.
+   */
+  public CompletableFuture<Void> runSave(final GameData gameData, final HistoryNode node, final File file) {
```
Comment (ssoloff) [inline src/games/strategy/triplea/ui/export/ScreenshotExporter.java]: Agreed -- another YAGNI anticipating someone would want to call `runSave()` (and `promptForFile()`) independently.

File: src/games/strategy/ui/SwingWorkerCompletionWaiter.java [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+   *
+   * @param window The window to display while the operation is in progress; must not be {@code null}.
+   */
+  public SwingWorkerCompletionWaiter(final Window window) {
+    this.window = checkNotNull(window);
+  }
+
+  @Override
+  public void propertyChange(final PropertyChangeEvent event) {
+    if ("state".equals(event.getPropertyName())) {
```
Comment (ssoloff) [inline src/games/strategy/ui/SwingWorkerCompletionWaiter.java]: "state" is the name of the property `SwingWorker` changes when the state of the background task changes.  Ideally, it would be defined as a constant on `SwingWorker`, but it's unfortunately not.  I'll extract a package-private constant in `SwingWorkerCompletionWaiter` so

- it avoids a magic number, and
- it can be reused by the corresponding test (which also uses the same literal string).

File: src/games/strategy/triplea/ui/export/ScreenshotExporter.java [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+import games.strategy.triplea.ui.TripleAFrame;
+import games.strategy.triplea.ui.mapdata.MapData;
+import games.strategy.ui.SwingComponents;
+import games.strategy.ui.Util;
+
+/**
+ * Exports a game screenshot.
+ */
+public final class ScreenshotExporter {
+  private final TripleAFrame frame;
```
Comment (ssoloff) [inline src/games/strategy/triplea/ui/export/ScreenshotExporter.java]: The `TripleAFrame` instance is used during the screenshot export for the following reasons:

1. To provide a parent component for various dialogs that are displayed.
1. To provide the `IUIContext` instance used to retrieve various screenshot properties (e.g. scale, title, color, etc.).
1. To provide the `MapPanel` instance used to draw the actual screenshot image.

(1) is easy to address by simply using a `null` parent component for all dialogs, which is what several methods in `SwingComponents` already do.  (Note that the TripleA code I've examined seems to inconsistently choose between displaying a dialog relative to an actual parent component or relative to the screen.  I've been choosing the former in this PR.)

(2) and (3) might require some thought.  One could simply inline the `IUIContext` and `MapPanel` as parameters, but I'm not sure if that's any better.  If not for (1), I probably would say inlining them is acceptable.  Otherwise, it seems that the `TripleAFrame` is simply being used as a sort of ad-hoc Parameter Object.

File: src/games/strategy/ui/SwingWorkerCompletionWaiter.java [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+  /**
+   * Initializes a new instance of the {@code SwingWorkerCompletionWaiter} class.
+   *
+   * @param window The window to display while the operation is in progress; must not be {@code null}.
+   */
+  public SwingWorkerCompletionWaiter(final Window window) {
+    this.window = IWindow.fromAwt(checkNotNull(window));
+  }
+
+  SwingWorkerCompletionWaiter(final IWindow window) {
```
Comment (DanVanAtta) [inline src/games/strategy/ui/SwingWorkerCompletionWaiter.java]: Please mark with `@VisibleForTesting`. It appears this would be private if it were not for tests.

File: src/games/strategy/ui/SwingWorkerCompletionWaiter.java [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+      if (SwingWorker.StateValue.STARTED == event.getNewValue()) {
+        window.setVisible(true);
+      } else if (SwingWorker.StateValue.DONE == event.getNewValue()) {
+        window.setVisible(false);
+        window.dispose();
+      }
+    }
+  }
+
+  static interface IWindow {
```
Comment (DanVanAtta) [inline src/games/strategy/ui/SwingWorkerCompletionWaiter.java]: Please avoid/remove the `I` prefix on interfaces. (http://www.triplea-game.org/dev_docs/dev/code_format - http://google.github.io/styleguide/javaguide.html#s5.1-identifier-names)

Doing stuff specifically for test only, can be a smell, yes. In this case you are allowing hard to test code to be testable. You lose perhaps one point for the test scaffolding, but gain perhaps about 5 for testable code. In another sense you are also decoupling the UI interaction API with the UI implementation. That is a very good thing and is worth perhaps 5 more "points."  :+1: 

Looking at the API and usage, there is not strict need to keep the exact Swing API. You could combine both the `dispose` and `setVisible` methods into perhaps a `close()` method, where you would call both.

The static method to do a default conversion I think is pretty clean. Overall this type of interface is probably a good pattern for a first iteration of getting some of the Swing code out of the game logic code.  So overall :+1:  from me here.

File: src/games/strategy/util/FileUtil.java [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+   * @param extension The extension to append; must not be {@code null}.
+   *
+   * @return The file, possibly with the extension appended; never {@code null}.
+   *
+   * @throws IllegalArgumentException If {@code extension} starts with a period.
+   */
+  public static File appendExtensionIfAbsent(final File file, final String extension) {
+    checkNotNull(file);
+    checkNotNull(extension);
+    checkArgument(!extension.startsWith(PERIOD), "extension must not start with a period");
```
Comment (DanVanAtta) [inline src/games/strategy/util/FileUtil.java]: There is a principle to be lenient on input, but strict on output. In this case that would mean we should try to accept either type of input, with leading period or not.

We can probably overcome this by having a pretty simple intermediate variable, eg:
```
String extensionWithPeriod = extension.startsWith(PERIOD) ? extension : PERIOD + extension
```

File: src/games/strategy/util/FileUtil.java [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+   * </p>
+   *
+   * @param file The file to which the extension will be appended; must not be {@code null}.
+   * @param extension The extension to append; must not be {@code null}.
+   *
+   * @return The file, possibly with the extension appended; never {@code null}.
+   *
+   * @throws IllegalArgumentException If {@code extension} starts with a period.
+   */
+  public static File appendExtensionIfAbsent(final File file, final String extension) {
```
Comment (DanVanAtta) [inline src/games/strategy/util/FileUtil.java]: Not sure. A FileUtil class may not be the best either. These can become dumping grounds for very unrelated methods. My first inclination is to just keep this with the class that uses it for now.

File: src/games/strategy/ui/SwingComponents.java [hunk 20]
```diff
@@ -278,4 +290,102 @@ public static JMenu newJMenu(final String menuTitle, final KeyboardCode keyboard
...
+   * @param<T> The type of the task result.
+   *
+   * @param frame The {@code Frame} from which the progress dialog is displayed or {@code null} to use a shared, hidden
+   *        frame as the owner of the progress dialog.
+   * @param message The message to display in the progress dialog; must not be {@code null}.
+   * @param task The task to be executed; must not be {@code null}.
+   *
+   * @return A promise that resolves to the result of the task; never {@code null}.
+   */
+  public static <T> CompletableFuture<T> runWithProgressBar(
```
Comment (ssoloff) [inline src/games/strategy/ui/SwingComponents.java]: Are you referring to the return value of `runWithProgressBar` or the return value of the `Callable` passed as the `task` parameter of `runWithProgressBar`?

If the former, then the answer is yes.  That is how we are able to chain the behavior to be run after the task is complete (i.e. via `whenComplete`).

If the latter, then it's a good point, and the answer is maybe.  Let me explain how it came about...

I originally added an overload for `runWithProgressBar` that accepted a `Runnable` for the case where no return value from the background task is needed (such as the screenshot export code under discussion):

    public static CompletableFuture<Void> runWithProgressBar(
            final Frame frame,
            final String message,
            final Runnable task) {
        return runWithProgressBar(frame, message, () -> {
            task.run();
            return null;
        });
    }

However, I ran into the classic problem of how to handle a checked exception thrown within the `Runnable` (specifically, in this case, because `ScreenshotExporter#runSave` may throw `IOException`).  I considered the following possible solutions to this problem, but none of them seemed satisfactory:

- tunnel the checked exception through an unchecked exception
- introduce a new interface (e.g. `ThrowingRunnable`) that mimics `Runnable` but allows a checked exception to be thrown from `run`
- using a [sneaky throw](http://www.philandstuff.com/2012/04/28/sneakily-throwing-checked-exceptions.html)

In the end, I simply chose to use a `Callable` so that methods that effectively return `void` are represented as `Callable<Void>`, with the ugly `return null` as you observed.

If you prefer one of the alternatives I listed above, or have another suggestion, I'm happy to change it.

**Comment by ssoloff** [comment]
The latest set of commits address issues raised in the [most recent review](https://github.com/triplea-game/triplea/pull/1419#pullrequestreview-15614922).  The only pending issue is the [question](https://github.com/triplea-game/triplea/pull/1419#discussion_r95069085) about the `runWithProgressBar()` return value, to which I provided a [response](https://github.com/triplea-game/triplea/pull/1419#discussion_r95072731).

Because we're (hopefully) near the end of this PR, I need to start thinking about squashing and rebasing these commits.  When you're ready to approve the PR, would it be acceptable to simply squash everything into a single commit?  My commits were rather fine-grained during the code review, so I can't see any logical chunks into which I can break them.

File: test/games/strategy/ui/SwingComponentsTest.java [hunk 21]
```diff
@@ -29,4 +31,20 @@ public void testAppendExtensionIfAbsent_ShouldHandleExtensionThatStartsWithPerio
...
+  public void testExtensionWithLeadingPeriod() {
+    assertThat(extensionWithLeadingPeriod("aaa"), is(".aaa"));
+    assertThat(extensionWithLeadingPeriod(".aaa"), is(".aaa"));
+    assertThat(extensionWithLeadingPeriod("aaa.aaa"), is(".aaa.aaa"));
+    assertThat(extensionWithLeadingPeriod(".aaa.aaa"), is(".aaa.aaa"));
+  }
+
+  @Test
+  public void testExtensionWithoutLeadingPeriod() {
+    assertThat(extensionWithoutLeadingPeriod("aaa"), is("aaa"));
```
Comment (DanVanAtta) [inline test/games/strategy/ui/SwingComponentsTest.java]: test cases for 2 or 1 character extensions would be interesting. NBD though.

Comment (ssoloff) [inline test/games/strategy/ui/SwingComponentsTest.java]: Done.  I also added a test case for the empty extension boundary condition (which didn't behave as I expected).  I'll proceed to rebase and squash -- it should be ready later today.

BTW, for my future reference, is there a preferred way in TripleA to write parameterized tests (other than doing what I did here by replicating the test multiple times)?  I did a quick search and didn't see anything obvious (e.g. the use of JUnit's `Parameterized` test runner).  I personally don't like the `Parameterized` runner that much, principally because of the need to extract one fixture per parameterized test.  I've used [JUnitParams](https://github.com/Pragmatists/JUnitParams) in the past and like it mostly because I can mix parameterized and non-parameterized tests in a single fixture.

**Comment by ssoloff** [comment]
@DanVanAtta I've rebased and squashed my branch.  I did some final manual testing to ensure everything's still working as expected, and it all looks good.  (I did discover a minor bug related to exporting a screenshot from the Game History panel, but it is present on `master`, as well, so I'll write up a separate issue for it.)

Please feel free to merge whenever it's convenient.  Thanks for being patient with me during the code review!

**Comment by DanVanAtta** [comment]
Was a pleasure, thank you for the submission
